### PR TITLE
Import Implied

### PR DESCRIPTION
**Pull request by odersky**
Introduce a new form `import implied` to import implied instances. Regular imports don't touch them anymore. See: 16cb974 for the docs.

### CODE REVIEW

**Comment by odersky** [comment]
I am counting on the bootstrap to test this in anger, both for functionality and for usability. Adding it now, since it is always better to start more restrictive instead of tightening the rules later.

File: tests/neg/import-implied.scala [hunk 1]
```diff
@@ -0,0 +1,29 @@
+class TC
+object A {
+  implied tc for TC
+  def foo given TC = ()
+}
+object B {
+  import A._
+  foo            // error: no implicit argument was found
+  foo given tc   // error: not found: tc
```
Comment (liufengyun) [inline tests/neg/import-implied.scala]: Forbidding explicit usage of `tc` here is a little counter-intuitive to lexical scoping, given the literal reading of `import A._`.

What about the following:

- `import A._` to import all members for _explicit_ usage
- `implied A._` to enable implied instances defined in `A` for _implicit_ resolution
- `implied A.c` make `A.c` available for implicit resolution, no matter _A.c_ is implicit or not.

The last one is up to debate, the idea is to encourage _use-site implicits_ over _definition-site implicits_.

Comment (arturopala) [reply tests/neg/import-implied.scala]: Would it be possible to use this feature to make any member a candidate for implicit resolution in scope, i.e. `import implied A.foo`?

**Comment by odersky** [comment]
@liufengyun What you propose is definitely worth considering. There's a complication however, in that the "implicitness" of a definition now depends not only on its definition, but also on the way it is imported. This is likely to have profound effects on spec and implementation. 

The current PR avoids this problem, and results in the narrowest possible import semantics. So I propose to get this in now since we can always generalize later.

File: compiler/src/dotty/tools/dotc/core/TypeErrors.scala [hunk 2]
```diff
@@ -18,7 +18,7 @@ import config.Printers.cyclicErrors
 class TypeError(msg: String) extends Exception(msg) {
   def this() = this("")
   def toMessage(implicit ctx: Context): Message = super.getMessage
-  override def getMessage: String = throw new Exception("Use toMessage instead for TypeError")
+  override def getMessage: String = super.getMessage
```
Comment (liufengyun) [inline compiler/src/dotty/tools/dotc/core/TypeErrors.scala]: The call `super.getMessage` will get empty string, which may hide usual error messages.

File: compiler/src/dotty/tools/dotc/core/tasty/TastyFormat.scala [hunk 3]
```diff
@@ -324,7 +325,8 @@ object TastyFormat {
   final val OPAQUE = 35
   final val EXTENSION = 36
   final val GIVEN = 37
-  final val PARAMsetter = 38
+  final val IMPLIED = 38
+  final val PARAMsetter = 39
```
Comment (liufengyun) [inline compiler/src/dotty/tools/dotc/core/tasty/TastyFormat.scala]: Advance tasty version, as the format changes?

**Pull request merged**